### ISSUE TITLE
release: adopt "keep-a-changelog" style in CHANGELOG.md

### ISSUE THREAD

**Issue by hyangah** [golang/vscode-go#3531]
Until now, we prepared the CHANGELOG entry during the stable version release process.
The CHANGELOG.md file is embedded in the released extension and is also used by VS Marketplace's extension info page
(https://marketplace.visualstudio.com/items/golang.Go/changelog).

The current way has problems:

  * We want the release process to involve fewer manual steps and the responsibility to be shared among team members. Writing the CHANGELOG.md makes this step challenging.
  * It's better to write the CHANGELOG entry when the notable features/changes/fixes are made. And, the contributor who made the change is the best person to write the note.
  * Pre-release extensions are built from the master, often before we write the CHANGELOG entry. That makes it hard for users to learn about notable changes they need to pay attention while using the pre-release versions.
  * VS Marketplace seems to use the CHANGELOG.md from the latest version of extension (regardless whether it is pre-release or stable release). That can be confusing.
  * The CHANGELOG.md sometimes needs fixes/clarification/edits after publication. In the current way, fixing errors in the CHANGELOG.md requires a new extension version publication, which is most likely unnecessary.
  
 To address the problems, we propose to reorganize how CHANGELOG.md file is maintained.
 
 * Adopt [Keep a changelog](https://keepachangelog.com/en/1.1.0/) style - which includes the "Unreleased" section. The section will list notable changes that are still in the pre-release phase.
 * Have the CHANGELOG.md in the root of the repo, so it is more visible from the repo.
 * Instead of embedding the full copy of the CHANGELOG.md file in the extension, have a static CHANGELOG.md that contains a link to the CHANGELOG.md file in the repo master. Typos and small edits can be made directly on the master and don't require extension publication.
 * Recommend contributors to have their notable changes to be documented with their changes.
 * The relui-based release automation process will move "Unreleased" entries to the released version section and get copied over to the GH Release page.
 * New features and setting changes result in a minor version bump. 
 * Patch releases should be strictly for fixes in regression and minor changes. CHANGELOG.md will be replaced with the link to the milestone instead.
 
 cc @h9jiang

**Comment by h9jiang** [comment]
One thing I noticed while developing the release automation is about the CHANGELOG heading .

Right now, the change log is organized based on version (which make sense) and ordered by dates (which also make sense). 

But one thing that imo is not making sense is including the release month/year in the heading. 

```
## v0.42.0 - 17 Jul, 2024

A comprehensive list of changes can be found in the complete [commit history](https://github.com/golang/vscode-go/compare/v0.41.4..v0.42.0).

### detail 1
### detail 2
### detail 3
```

For example: 

https://github.com/golang/vscode-go/blob/master/extension/CHANGELOG.md#v0420

This link does not take you to anywhere because this link does not link to v0.42.0 entry. Instead, you should add the release month and year. 

https://github.com/golang/vscode-go/blob/master/extension/CHANGELOG.md#v0420---17-jul-2024



I believe having the month and year is useful information but should not be included in the title.

**Comment by hyangah** [comment]
Agree. From the new releases, let's use this format from new releases.

```
### v0.44.0

Date: 2024-10-01

....
```
